async fx compile

aorenste opened the pull request:
Adds the ability to run the selected out-of-process fx compile scheme in async mode - where we kick off the compile and then run eagerly until the compile is finished.

Added a test which runs a tiny model in a loop making sure that we execute it both eagerly and then compiled.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #146135



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @desertfire @chauhang @aakhundov

Differential Revision: [D71135546](https://our.internmc.facebook.com/intern/diff/D71135546)

aorenste committed:
WIP: async compile

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste committed:
Update on "WIP: async compile"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71135546).

jamesjwu commented on test/inductor/test_compile_subprocess.py:
```diff
@@ -66,6 +67,64 @@ def tearDown(self):
+            stack.enter_context(patch(f"{inductor_config}.autotune_local_cache", False))
+            stack.enter_context(patch(f"{inductor_config}.fx_graph_cache", False))
+            stack.enter_context(patch(f"{func_config}.enable_autograd_cache", False))
+            stack.enter_context(patch(f"{func_config}.strict_autograd_cache", True))
```
nit: if you're disabling autograd cache you probably don't need this line after

jamesjwu commented on torch/_inductor/compile_fx_ext.py:
```diff
@@ -532,6 +552,7 @@ def _run_in_child(
 
 # This is a debugging/testing implementation of FxCompile which serializes the
 # input and output but still runs the FxCompile in-process.
+@final
```
Why's this class final now?

aorenste replied:
Just documentation that it doesn't expect anything to subclass it.

jamesjwu commented on torch/_inductor/compile_fx_async.py:
```diff
@@ -0,0 +1,169 @@
+        self._callback = callback
+
+        # This tells callers to call us with `inputs` instead of `*inputs`.
+        self._boxed_call = True
```
I know that all inductor generated artifacts will always return with _boxed_call = True today, but shouldn't we check whether the produced compiled_artifact has _boxed_call = True, and thread that value through here?

jamesjwu reviewed (approved):
Seems fine to me

jansel commented on torch/_inductor/compile_fx_async.py:
```diff
@@ -0,0 +1,169 @@
+    # Number of times we ran "eager"
+    _stat_eager_runs: int = 0
+    # Number of times we ran our compiled (out-of-process) artifact
+    _stat_oop_runs: int = 0
```
```suggestion
    _stat_compiled_runs: int = 0
```
The name `oop` was confusing me.  I think out-of-process is implied by the class we are in.

aorenste commented on test/inductor/test_compile_subprocess.py:
```diff
@@ -66,6 +67,64 @@ def tearDown(self):
+                # DEBUGGING: Print a periodic message so we know we're still
+                # running...
+                now = time.time()
+                if TICK_REPORT is not None and (now - last_report > TICK_REPORT):
```
Here we don't have any counter to track - for debugging I just wanted something to indicate that we weren't hung.

aorenste committed:
Update on "async fx compile"

Adds the ability to run the selected out-of-process fx compile scheme in async mode - where we kick off the compile and then run eagerly until the compile is finished.

Added a test which runs a tiny model in a loop making sure that we execute it both eagerly and then compiled.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 amjames desertfire chauhang aakhundov

Differential Revision: [D71135546](https://our.internmc.facebook.com/intern/diff/D71135546)

[ghstack-poisoned]

aorenste commented:
@aorenste has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D71135546).
